src/librustc_data_structures/graph/dominators/mod.rs

```diff
@@ -159,6 +160,36 @@ impl<'dom, Node: Idx> Iterator for Iter<'dom, Node> {
     }
 }
 
+pub struct DominatorFrontiers<G: ControlFlowGraph> {
+    dfs: IndexVec<G::Node, BitSet<G::Node>>,
+}
+
+impl<G: ControlFlowGraph> DominatorFrontiers<G> {
+    pub fn new(graph: &G, doms: &Dominators<G::Node>) -> Self {
+        let num_nodes = graph.num_nodes();
+        let mut dfs = IndexVec::from_elem_n(BitSet::new_empty(num_nodes), num_nodes);
+
+        for b in (0..num_nodes).map(|i| G::Node::new(i)) {
+            if graph.predecessors(b).take(2).count() > 1 {
```

Comment:
What's the maximum number of predecessors? Or is there no upper bound?

Reply:
Here we only care if it's a join point -- i.e. it has more than one predecessor.

In general there can be unbounded predecessors.

As you can see from a later comment, this code isn't as nice as I'd hoped, but hey.

Reply:
![image](https://user-images.githubusercontent.com/604955/55716077-ff7cc900-59ed-11e9-9cec-8a2c387edde5.png)

Reply:
I was partly wondering if there's a nicer way of saying "how many predecessors does `b` have"? rather than firing up an iterator. It might not be worth worrying about (depending on how expensive the iterator is to set up).

Reply:
Yeah, I was hoping for this too!

Predecessors are expressed by a trait with a sole method which returns an iterator:
https://github.com/rust-lang/rust/blob/3750348daff89741e3153e0e120aa70a45ff5b68/src/librustc_data_structures/graph/mod.rs#L35-L43

And then the MIR implementation of this trait is in terms of `predecessors_for()`:
https://github.com/rust-lang/rust/blob/3750348daff89741e3153e0e120aa70a45ff5b68/src/librustc/mir/mod.rs#L233-L236

Let me have a look if I can use any of this stuff directly, but I might be hard to access as we are behind parameterised types.

Reply:
[Another option is to `collect()` the predecessors]

Reply:
Yeah, so at the point where we are fiddling with those iterators, we are working with a `ControlFlowGraph`, which a `MIR` is an instance of. We can't get directly at the underlying `predecessors_for()` of the `MIR` struct, I'm afraid.

Reply:
OK, then I think this is as good as it gets. Using `collect` would almost certainly be worse than what we have.

Reply:
We *could* implement this function (elsewhere, probably in `emit_tir.rs`) as a concrete function over a MIR of course. I only implemented it here because the rest of the code from that paper was also here and already generic.

Reply:
It's not worth it. I just wanted to check if something easy had been missed.

Reply:
OK.